### PR TITLE
Enable typescript-eslint

### LINKED ISSUE #5694: Enabling TypeScript support in eslint
### Discord username

_No response_

### How important is this feature to you?

3 – Lack of it makes using Enso slightly harder

### Describe the idea you'd like to see implemented.

This issue should be implemented as a PR to this branch: https://github.com/enso-org/enso/tree/wip/wdanilo/shader-compilation-improvement-184304289

It will be later moved to develop when this branch will be closed.

Right now, we have a few issues:
1. We have multiple eslint configurations instead of one - you can find them in `app/ide-desktop/lib/client` and `app/ide-desktop/lib/content`.
2. The eslint config has many lines commented that require TypeScript type checker support. When you uncomment these lines, there is an eslint error shown that needs to be fixed. What's interesting, we are successfully using eslint with these lines uncommented in another place: `lib/rust/ensogl/pack/src`.

The goal of this task is to make eslint working correctly with TS types support and unifying the configurations preferably to one file.

### Is your feature request related to a problem?

_No response_

### Screenshots, screencasts, mockups.

_No response_

### Would you be willing to help us implement this feature?

No

### PR DESCRIPTION
**Pull request by Nctdt**
### Pull Request Description
implement [issue#5694](https://github.com/enso-org/enso/issues/5694)

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

**Review by wdanilo — CHANGES_REQUESTED** [review]
After these changes, the app does not compile anymore. After running `./run ide build --skip-version-check --skip-wasm-opt --backend-source release --backend-release latest` I see a lot of errors like:

```
INFO ide_ci::program::command: npmℹ️ ../client/src/bin/project-manager.ts(4,25): error TS2307: Cannot find module 'config' or its corresponding type declarations.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config.ts(6,25): error TS2307: Cannot find module 'naming' or its corresponding type declarations.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config.ts(7,24): error TS2307: Cannot find module 'paths' or its corresponding type declarations.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(5,25): error TS2307: Cannot find module 'config' or its corresponding type declarations.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(7,25): error TS2307: Cannot find module 'naming' or its corresponding type declarations.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(62,31): error TS18046: 'group' is of type 'unknown'.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(63,30): error TS18046: 'group' is of type 'unknown'.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(74,13): error TS18046: 'option' is of type 'unknown'.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(78,38): error TS18046: 'option' is of type 'unknown'.
 INFO ide_ci::program::command: npmℹ️ ../client/src/config/parser.ts(248,85): error TS7006: Parameter 'option' implicitly has an 'any' type.
 INFO ide_ci::program::command: npmℹ️ ../client/src/index.ts(8,25): error TS2307: Cannot find module 'config' or its corresponding type declarations.
```

I've noticed that you removed a bunch of `tsconfig` files from some libs. This is probably the cause of that. Why did you remove them? These are separate packages and they should have tsconfig files.

**Comment by Nctdt** [comment]
Currently have some problems that need to resolve.

**Comment by wdanilo** [comment]
Ok, @Nctdt, please ping me when it will be resolved. I would love to merge it as soon as possible :) After you fix it, I will do some additional review / some additional fixes to docs and merge it right away

**Comment by somebody1234** [comment]
hey @wdanilo - just wondering if adding more lints would be in-scope for this PR?

a lot of them would just be for consistency, and so they may be too strict - but on the upside, that means they'd be able to be autofixed

```js
        'no-restricted-syntax': [
          'error',
          {
            'selector': ':matches(ImportSpecifier, ExportDeclaration, ExportSpecifier)',
            'message': 'No {} imports and exports'
          }
        ],
        'no-restricted-syntax': [
          'error',
          {
            'selector': ':matches(ExportAllDeclaration)',
            'message': 'No re-exports'
          }
        ],
        // prefer interface over type
        '@typescript-eslint/consistent-type-definitions': 'error',
        '@typescript-eslint/consistent-type-imports': [
            'error',
            { prefer: 'no-type-imports' },
        ],
        '@typescript-eslint/member-ordering': 'error',
        // Method syntax is not type-safe.
        // See: https://typescript-eslint.io/rules/method-signature-style
        '@typescript-eslint/method-signature-style': 'error',
        '@typescript-eslint/naming-convention': [
            'error',
            {
                selector: 'function',
                // PascalCase for React components, camelCase for all other situations
                format: ['camelCase', 'PascalCase'],
            },
        ],
        '@typescript-eslint/no-confusing-void-expression': 'error',
        // React 17 and later supports async functions as event handlers, so we need to disable this
        // rule to avoid false positives.
        //
        // See: https://github.com/typescript-eslint/typescript-eslint/pull/4623
        '@typescript-eslint/no-misused-promises': [
            'error',
            { checksVoidReturn: { attributes: false } },
        ],
        '@typescript-eslint/no-redundant-type-constituents': 'error',
        '@typescript-eslint/no-type-alias': [
            'error',
            { allowAliases: 'in-unions-and-intersections', allowLiterals: 'in-intersections' },
        ],
        '@typescript-eslint/no-unnecessary-condition': 'error',
        '@typescript-eslint/no-useless-empty-export': 'error',
        '@typescript-eslint/parameter-properties': [
            'error',
            { prefer: 'class-property' },
        ],
        '@typescript-eslint/prefer-enum-initializers': 'error',
        '@typescript-eslint/prefer-readonly': 'error',
        '@typescript-eslint/require-array-sort-compare': 'error',
        '@typescript-eslint/restrict-template-expressions': 'error',
        '@typescript-eslint/sort-type-constituents': 'error',
        '@typescript-eslint/switch-exhaustiveness-check': 'error',
        'default-param-last': 'off',
        '@typescript-eslint/default-param-last': 'error',
        'no-invalid-this': 'off',
        '@typescript-eslint/no-invalid-this': [
            'error',
            { capIsConstructor: false },
        ],
        'jsdoc/no-magic-numbers': 'off',
        '@typescript-eslint/no-magic-numbers': 'error',
        'no-redeclare': 'off',
        // Important to warn on accidental duplicated `interface`s e.g. when writing API wrappers
        '@typescript-eslint/no-redeclare': 'error',
        'no-shadow': 'off',
        '@typescript-eslint/no-shadow': 'warn',
        'no-unused-expressions': 'off',
        '@typescript-eslint/no-unused-expressions': 'error',
```

**Comment by wdanilo** [comment]
@somebody1234 absolutely, feel free to add/remove any lints that you feel would make the codebase safer/better :)

**Comment by somebody1234** [comment]
the [ts-eslint-2](https://github.com/enso-org/enso/tree/wip/somebody1234/ts-eslint-2.) branch is now ready to test, if anyone wants to do that.

i've added a bunch of `no-restricted-syntax` rules to ensure more consistent naming and code style - please tell me if:
- i went overboard
- the convention should be switched to another
- they should be commented better

there are a couple things i wanted to add but left out because they would require significant changes, and i'm not sure what the 
- `require-jsdoc` is currently set to default instead of on for everything - almost everything would show a warning if this was fully enabled.
- it would be nice to be consistent between function declarations `function foo() {}` vs arrow functions `const foo = () =>` for top-level functions. right now most of them seem to be `function foo() {}` but some of the new PRs might introduce some arrow functions. note that which one is chosen will make no difference for typechecking. using arrow functions will lock us out of using overloads though, which we *may* want to use down the line. @wdanilo thoughts?

**Comment by somebody1234** [comment]
not sure what's the best course of action - should i merge ts-eslint-2 into this one? create a new PR so that it can be reviewed separately?

**Comment by wdanilo** [comment]
> not sure what's the best course of action - should i merge ts-eslint-2 into this one? create a new PR so that it can be reviewed separately?

Let's do whatever works best for you / whatever is easiest. We just need a single PR for a fast review, it can be this one, or this can be closed and we can have another one, up to you :)

Regarding the questions:
- `please tell me if [...] i went overboard [...]` - knowing your code quality, its probably perfect, but create a single PR and I'd love to give it a fast look.
- `require-jsdoc`- I feel that enabling it for everything might be an overkill. TBH, I feel that the LESS docs we have and the more self-documented code we have (good variable names, nicely refactored, short functions), the better. Sometimes it is impossible (because of code complexity or delivery time constraints), so docs are needed evil. Requiring it everywhere would make the code more cluttered and hard to read. What do you think?
- Regarding top-level functions, I'm both hands for `function ...` syntax, mainly because go overloads and easy of reading the code (you immediately see what it is without the need of parsing with your eyes more code). The former point is more important though.

**Comment by somebody1234** [comment]
- `require-jsdoc` totally agree
- function syntax - sounds great. not sure what you mean by the former point - you mean the one about consistency?

**Comment by wdanilo** [comment]
> * `require-jsdoc` totally agree
> * function syntax - sounds great. not sure what you mean by the former point - you mean the one about consistency?

@somebody1234 perfect :)

Regarding top-level functions, in my reply I provided two points: `I'm both hands for function ... syntax, mainly because [POINT 1] of overloads and [POINT 2] easy of reading the code (you immediately see what it is without the need of parsing with your eyes more code). The former point is more important though.`

Former, means POINT 1 here. Sorry for not being clear in my message!

**Comment by somebody1234** [comment]
should be ready for review now.
[here](https://github.com/enso-org/enso/pull/5717/commits/91e9b055202ddfb9fa0b52e02c0a9133ff184423) are the changes from the merge
[here](https://github.com/enso-org/enso/pull/5717/commits/941323e33cd6fca4664575d7aa23b24aff6e53d7) are the changes from the latest commit

i'll run an ide build real quick just to make sure things are working - please let me know if there's anything else i should be doing to test this

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js]: What do you mean by "The .d.ts is wrong"? I do not get these comments.

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js]: Here is a note telling that something has to be enabled without explanation why it has toi be.

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js:9]: Why we need this?

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js:15]: Why we need this?

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js:28]: This looks like something extremely painful to maintain. Is there any other way or we need to list these things explicitly here? :(

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [inline app/ide-desktop/eslint.config.js:355]: Oh wow, this is the biggest linter config I've seen ever :D Ehh, JavaScript world :P

File: app/ide-desktop/lib/client/src/bin/server.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/bin/server.ts]: TBH, I much more preferred the older `logger.error` over `contentConfig.LOGGER` syntax. The new one is super verbose. Can we keep the old one, please? If you prefer the new one, could you explain why, please?

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/config/parser.ts:79]: Why this change? I like `== null` more as it tests for both `null` nad `undefined`, which is more secure for possibly future code refactoring. This comment applies to other similar places. If you feel the new form is better, I'd be thankful for short explanation of why :)

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/config/parser.ts:101]: Why this change? Refactoring the literal to entry made the code shorter and less repetitive.

File: app/ide-desktop/lib/client/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/index.ts]: This is not a constant in the "traditional" meaning. This is just a variable pointing to a class instance, so IMO this should be written `logger`, just like `console` is `console` and not `CONSOLE`.

File: app/ide-desktop/utils.ts
Comment (wdanilo) [inline app/ide-desktop/utils.ts]: Fixme left in the code.

**Review by wdanilo — APPROVED** [review]
Oh wow, you added a LOT of additional changes. All of them are really nice, thanks so much for keeping it high quality! I added a few comments, none of them would block this, thus accepting! :)

File: app/ide-desktop/eslint.config.js
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:9]: i've added lints to prefer namespace imports: `* as foo`. there is an explicit whitelist elsewhere but i don't think it needs to cover modules that are only used here. i can almost certainly make it clearer though

File: app/ide-desktop/eslint.config.js
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:15]: this one is needed to re-enable the lint disabled on line 10, otherwise the rest of the document won't be linted

File: app/ide-desktop/eslint.config.js
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js:28]: it might be possible to programmatically list directories and generate this string - would that be preferable?

File: app/ide-desktop/eslint.config.js
Comment (somebody1234) [reply app/ide-desktop/eslint.config.js]: the typescript definitions provided by the modules don't error when importing as namespace imports. i don't think it's important to mention this though, so might as well remove this comment

File: app/ide-desktop/lib/client/src/bin/server.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/bin/server.ts]: this was just auto-refactored because we were importing as `{ logger }` before. i can extract it to a const so it will be `LOGGER.error` - would that work?

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/config/parser.ts:101]: i removed `entry` since it requires either:
- a type annotation (longer, adds a tiny bit of maintenance cost), or
- `as const` - this is what it was previously, i just figured it'd be good to minimize usage of `as` (even `as const`) because `as T` (not `as const`) isn't quite type-safe.

thoughts?

tl;dr: refactoring to `entry` *is* shorter and less repetitive, but i think avoiding explicit type annotations is a worthwhile tradeoff

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/src/config/parser.ts:79]: it should be safe to use `=== undefined` since typescript will error if it's incorrect:
- `null` and `{ any object type } | undefined` have nothing in common
- similar reasoning for accidentally comparing a `T | null` (from `RegExp.match` or `document.getElementById`) against `undefined`

as for why disallow `null`, it's mostly personal preference - plus afaict most if not all of the APIs returning `null`  are legacy APIs.

i have nothing against `== null` though, and i'd be happy to switch back. the reason why i chose to do things this way is because of the `eqeqeq` rule that prefers `===` over `==` - but it has an option to prefer `== null` so it'd be *very* simple to change back.

**Comment by somebody1234** [comment]
i'm considering adding a new lint to disallow `val as T`, with a whitelist for branded types `string & { _brand: "ProjectId"; }`. but that may be better left until brands are actually in the codebase

(edit: there are 4 places where ` as ` is used as of the latest commit, and it shouldn't hurt to `eslint-disable` them (with an explanation of course) if this new lint is added)

**Comment by somebody1234** [comment]
force-pushed after running prettier on the latest commit

**Comment by somebody1234** [comment]
should pretty much be ready for merge - another pair of eyes would be great though.

there are a couple of issues that are still waiting on confirmation:
`=== undefined` vs `== null`

consider automatically generating `const RELATIVE_MODULES` (needed to ensure imports are in the right order). i do agree it may be painful to maintain, but i'm not sure whether crawling directories inside `eslint.config.js` is a good idea.

File: app/ide-desktop/eslint.config.js
Comment (wdanilo) [reply app/ide-desktop/eslint.config.js:28]: Probably not now. But I'm sad it's so complex and that eslint can't do it by itself.

File: app/ide-desktop/lib/client/src/bin/server.ts
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/bin/server.ts]: As I've written below, I think it should be `logger.error`, as we are not writing `CONSOLE.log`. Do you agree?

Comment (somebody1234) [reply app/ide-desktop/lib/client/src/bin/server.ts]: yup - it's already been changed to `logger.error`

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/config/parser.ts:79]: `plus afaict most if not all of the APIs returning null are legacy APIs.` - oh, this is something new to me. If you want to return lack of value, what should you return? I'm never returning `undefined` in my APIs, as I learned that `undefined` means lack of definition in JS world, and it should never be used as a real value. 

After your explanation I'm 100% OK with this version, thanks for it. Le]ts keep it this way, sounds more "rigid" than my proposal :)

File: app/ide-desktop/lib/client/src/config/parser.ts
Comment (wdanilo) [reply app/ide-desktop/lib/client/src/config/parser.ts:101]: Makes sense. Ok, its up to you, whatever you like better :)

**Comment by wdanilo** [comment]
> i'm considering adding a new lint to disallow `val as T`, with a whitelist for branded types `string & { _brand: "ProjectId"; }`. but that may be better left until brands are actually in the codebase
> 
> (edit: there are 4 places where `as` is used as of the latest commit, and it shouldn't hurt to `eslint-disable` them (with an explanation of course) if this new lint is added)

Yes, let's do it. `as` is very dangerous.

**Comment by wdanilo** [comment]
> should pretty much be ready for merge - another pair of eyes would be great though.
> 
> there are a couple of issues that are still waiting on confirmation: `=== undefined` vs `== null`
> 
> consider automatically generating `const RELATIVE_MODULES` (needed to ensure imports are in the right order). i do agree it may be painful to maintain, but i'm not sure whether crawling directories inside `eslint.config.js` is a good idea.

I believe we are done with these for now :)

**Comment by wdanilo** [comment]
I would only ask for the final check that all things like `contentConfig.LOGGER` are changed to `logger` and let's merge it. If you've tested it and works, I believe we can just progress with merging :)

**Comment by somebody1234** [comment]
ok to summarize:
- autogenerating `RELATIVE_MODULES`: not worth the complexity, but may want to add a note 
- undefined vs null, [this](https://github.com/enso-org/enso/pull/5865#discussion_r1136475391) seems to summarize our philosophy. might want to investigate whether we should change which lints we use, or whether we should just switch to use `undefined` everywhere?
- `contentConfig.LOGGER` have all been changed to `logger` - this has been checked with multiple `grep`s
- `as` should be disallowed - might as well do that here (this should potentially be relaxed in future PRs that introduce branded types)

**Comment by wdanilo** [comment]
@somebody1234 One explanation regarding `undefined` vs `null`. Do you agree with this philosophy, or in TypeScript `undefined` is considered as the preferred way to indicate lack of value and `null` is used only ion obsolete APIs? I might have been missing the "newest trends of API design in TS". :) 

Regarding everything else, we are on the same page! :)

**Comment by somebody1234** [comment]
re: `undefined` vs `null`. i'm not quite sure about current trends...
but it is true that `null` *could* be used to indicate an explicit lack of value, however i think it's only useful when you need to handle `null` and `undefined` separately.

**Comment by wdanilo** [comment]
Ok, then let's just use `null` when creating the value explicitly and lets keep `undefined` as a special thing which means "undefined value", never created manually :) If you are OK with it, let's keep this rule here.

**Comment by somebody1234** [comment]
oh no, i still need to fix lints for `null`/`undefined`, and add a lint for `as`. should i revert the merge for now?

**Pull request merged**
